File: algorithms/ar-cochprb/java/coin_change_problem.java
class coinchange{ 
      
    // Returns the count of ways we can  
    // sum S[0...m-1] coins to get sum n 
   static long countWays(int S[], int m, int n) 
    { 
        //Time complexity of this function: O(mn) 
        //Space Complexity of this function: O(n) 
  
        // table[i] will be storing the number of solutions 
        // for value i. We need n+1 rows as the table is 
        // constructed in bottom up manner using the base 
        // case (n = 0) 
        long[] table = new long[n+1]; 
  
        // Initialize all table values as 0 
        Arrays.fill(table, 0);   //O(n) 
  
        // Base case (If given value is 0) 
        table[0] = 1; 
  
        // Pick all coins one by one and update the table[] 
        // values after the index greater than or equal to 
        // the value of the picked coin 
        for (int i=0; i<m; i++) 
            for (int j=S[i]; j<=n; j++) 
                table[j] += table[j-S[i]]; 
  
        return table[n]; }
      
    // Driver program to test above function 
    public static void main(String[] args) 
    { Scanner sc=new Scanner(System.in);
    int t=sc.nextInt();
    while(t-->0)
    {int m = sc.nextInt();
        int arr[] =new int[m];
        for(int i=0;i<m;i++)
        {
            arr[i]=sc.nextInt();
        }
        int k=sc.nextInt();
         
        System.out.println( countWays(arr, m, k)); 
    
        
    }
          
    } 
  
} 

File: algorithms/ar-cochprb/csharp/CoinChangeProblem.cs
using System;

public static class CoinChange
{
    public static int DoCoinChangeWithDynamicProgramming(int[] coins, int totalAmount)
    {
        // This array will finally contain the number of possibilities for an amount i at index i.
        var possibilities = new int[totalAmount + 1];

        // For an amount of 0 there is only one possibility:
        // Take no coin at all.
        possibilities[0] = 1;

        // Taking one coin at a time brings new possibilities for certain amounts.
        foreach (int coin in coins)
        {
            // In fact only for those amounts greater or equal to the coin itself.
            for (int amount = coin; amount <= totalAmount; amount++)
            {
                possibilities[amount] += possibilities[amount - coin];
            }
        }

        return possibilities[totalAmount];
    }

    public static void Main()
    {
        // var coins = new [] {1,2};
        // var amount = 3;
        // // => possibilites = 2;

        var coins = new [] {6,2,3};
        var amount = 8;
        // => possibilites = 3;
        
        int possibilities = DoCoinChangeWithDynamicProgramming(coins, amount);
        
        // Print the value here
        // Console.WriteLine($"There are {possibilities} possibilities");
    }
}